cluster: Add support to cluster to work with `NODE_OPTIONS="--inspect"

sameer-coder opened the pull request:
When using cluster and --inspect as cli argument it is correctly
handled and each worker will use a different port, this was
fixed by #13619. But when env var NODE_OPTIONS="--inspect"
is set this logic doesn't apply and the workers will fail as they
try to attach to the same port.

Fixes: https://github.com/nodejs/node/issues/19026

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x ] tests and/or benchmarks are included
- [x ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

#### Affected core subsystem(s)
- cluster

benjamingr commented:
Thanks for doing this!

@nodejs/v8-inspector  @nodejs/cluster

richardlau commented on lib/internal/cluster/master.js:
```diff
@@ -102,11 +102,14 @@ function createWorkerProcess(id, env) {
   const execArgv = cluster.settings.execArgv.slice();
   const debugArgRegex = /--inspect(?:-brk|-port)?|--debug-port/;
+  const nodeOptions = process.env.NODE_OPTIONS ?
+    process.env.NODE_OPTIONS.split(' ') : [];
```
There's no need to split `NODE_OPTIONS`. Later on `match()` can be called on the whole string.

sameer-coder replied:
@richardlau I have removed the split and doing a direct match. Please review again.

richardlau commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,27 @@
+
+if (process.config.variables.node_without_node_options)
+  common.skip('missing NODE_OPTIONS support');
+
```
Could you also call `common.skipIfInspectorDisabled()`?

sameer-coder replied:
Done

sameer-coder commented:
@richardlau Thanks for your comments. I am re-writing the test to spawn more workers like how its done in [test-inspector-port-cluster.js](https://github.com/mutantcornholio/node/blob/5e101d98aa253e7315d5f6caaa4796fe3aebfa09/test/inspector/test-inspector-port-cluster.js) test. Will get back to you.

cjihrig commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,27 @@
+    cluster.fork(env);
+
+    cluster.on('online', (worker) => {
+      process.exit(0);
```
Please don't call `process.exit()`. Instead, disconnect the worker and let the test exit naturally.

cjihrig commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,27 @@
+
+function checkForInspectSupport(flag) {
+  const env = Object.assign({}, process.env, { NODE_OPTIONS: flag });
+  const o = JSON.stringify(flag);
```
Could you use a more meaningful name than `o`?

sameer-coder commented:
@richardlau - I have updated the code to match the entire options string and also added common.skipIfInspectorDisabled().  Also, I have changed the test so that it spawns multiple workers now.

@cjihrig - I have made changes as per your comments to use worker disconnect instead of process.exit()

jasnell commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -2,26 +2,31 @@
     cluster.on('exit', (worker, code, signal) => {
-      assert.fail(`For ${o}, failed to start a cluster with inspect option`);
+      if (worker.exitedAfterDisconnect === false) {
+        assert.fail(`For ${nodeOptions}, failed to start cluster\
```
Nit: please avoid using multiline template strings.

sameer-coder replied:
@jasnell I have changed the code. No longer using multiline string. Please review again.

sameer-coder commented:
Did anyone get a chance to review this again?

benjamingr commented:
Added the author-ready tag - should land soon :)

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13663/

richardlau commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,31 @@
+const common = require('../common');
+const cluster = require('cluster');
+const assert = require('assert');
+const numCPUs = require('os').cpus().length;
```
To clarify, by "this" I mean the number of workers spawned in the subsequent for loop.

sameer-coder replied:
@richardlau you are right. I am no longer using numCPUs and directly spawning multiple workers.

richardlau commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,30 @@
+    });
+
+    cluster.on('exit', (worker, code, signal) => {
+      if (worker.exitedAfterDisconnect === false) {
```
Should this be an assert rather than if-fail?

sameer-coder replied:
Does this look ok?
```js
    cluster.on('exit', common.mustCall((worker, code, signal) => {
      assert(worker.exitedAfterDisconnect === true, `For NODE_OPTIONS failed to start cluster`);
    }, 2));
```

richardlau commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,30 @@
+      worker.disconnect();
+    });
+
+    cluster.on('exit', (worker, code, signal) => {
```
Wrap the callback in `common.mustCall`.

sameer-coder replied:
Done

richardlau commented on test/parallel/test-inspect-support-for-node_options.js:
```diff
@@ -0,0 +1,29 @@
+    cluster.on('exit', common.mustCall((worker, code, signal) => {
+      const errMsg = `For NODE_OPTIONS ${nodeOptions}, failed to start cluster`;
+      assert.strictEqual(worker.exitedAfterDisconnect, true, errMsg);
+    }, 2));
```
Tiny nit: maybe make the `2` here and in the for loop that forks the workers a const since they should match?

sameer-coder replied:
@richardlau I have added the const and replaced the `2` both in the for loop and callback.

richardlau commented:
One more CI: https://ci.nodejs.org/job/node-test-pull-request/13716/

richardlau commented:
Windows failure is https://github.com/nodejs/node/issues/19263 and unrelated to this PR.

richardlau commented:
Landed in https://github.com/nodejs/node/commit/9b34ea6161a637d6be7a68b816df8f8479d7c4ee.

richardlau commented:
Congratulations @sameer-coder on your first contribution to Node.js. 🎉

sameer-coder commented:
Thank you very much @richardlau for the opportunity to pick this up and then helping with the code.
I am absolutely thrilled at my first contribution and look forward to contributing more!
